src/OrleansEventSourcing/Common/NotificationTracker.cs

```diff
@@ -0,0 +1,226 @@
+        {
+            this.services = services;
+            this.listener = listener;
+            sendWorkers = new Dictionary<string, NotificationWorker>();
+            this.maxNotificationBatchSize = maxNotificationBatchSize;
+
+            foreach (var x in remoteInstances)
+            {
+                services.Verbose("Now sending notifications to {0}", x);
+                sendWorkers.Add(x, new NotificationWorker(this, x));
+            }
+        }
+
+        public void BroadcastNotification(INotificationMessage msg, string exclude = null)
+        {
+            foreach (var kvp in sendWorkers)
+            {
+                if (kvp.Key != exclude)
+                {
+                    var w = kvp.Value;
+                    w.Enqueue(msg);
+                }
+            }
+        }
+
+        /// <summary>
+        /// returns unresolved connection issues observed by the workers
+        /// </summary>
+        public IEnumerable<ConnectionIssue> UnresolvedConnectionIssues
+        {
+            get
+            {
+                return sendWorkers.Values.Select(sw => sw.LastConnectionIssue.Issue).Where(i => i != null);
+            }
+        }
+
+        /// <summary>
+        /// Update the multicluster configuration (change who to send notifications to)
+        /// </summary>
+        public void UpdateNotificationTargets(IReadOnlyList<string> remoteInstances)
+        {
+            var removed = sendWorkers.Keys.Except(remoteInstances);
+            foreach (var x in removed)
+            {
+                services.Verbose("No longer sending notifications to {0}", x);
+                sendWorkers[x].Done = true;
+                sendWorkers.Remove(x);
+            }
+
+            var added = remoteInstances.Except(sendWorkers.Keys);
+            foreach (var x in added)
+            {
+                if (x != services.MyClusterId)
+                {
+                    services.Verbose("Now sending notifications to {0}", x);
+                    sendWorkers.Add(x, new NotificationWorker(this, x));
+                }
+            }
+        }
+
+
+        public enum NotificationQueueState
+        {
+            Empty,
+            Single,
+            Batch,
+            VersionOnly
+        }
+
+        /// <summary>
+        /// Asynchronous batch worker that sends notfications to a particular cluster.
+        /// </summary>
+        public class NotificationWorker : BatchWorker
+        {
+            private NotificationTracker tracker;
+            private string clusterId;
+
+            /// <summary>
+            /// Queue messages
+            /// </summary>
+            public INotificationMessage QueuedMessage = null;
+            /// <summary>
+            /// Queue state
+            /// </summary>
+            public NotificationQueueState QueueState = NotificationQueueState.Empty;
```

Comment:
This is marked as public, but I'm not convinced it should be - at least not for writes.